docs/cookbook/prompt_engineering/text_based/self_ask.md

```diff
@@ -0,0 +1,248 @@
+            Follow up: Which player scored the first touchdown in Superbowl 1?
+            Intermediate answer: Max McGee.
+            Follow up: Which sports team did Max McGee play for?
+            Intermediate answer: Max McGee played for the Green Bay Packers.
+            So the final answer is: Green Bay Packers.
+            """
+        ),
+    ),
+]
+```
+
+Finally, we can use our Self-Ask implementation:
+
+```python
+response = self_ask_query(
+    query="The birth country of Jayantha Ketagoda left the British Empire when?",
+    examples=few_shot_examples,
+)
+print(response.content)
+# > Are follow up questions needed here: Yes.
+#   Follow up: What is the birth country of Jayantha Ketagoda?
+#   Intermediate answer: Jayantha Ketagoda is from Sri Lanka.
+#   Follow up: When did Sri Lanka leave the British Empire?
+#   Intermediate answer: Sri Lanka, formerly known as Ceylon, gained independence from the British Empire on February 4, 1948.
+#   So the final answer is: February 4, 1948.
+```
+
+## Enhancing Self-Ask Implementation
+
+To further improve our Self-Ask implementation, consider the following enhancements:
+
+- Dynamic example selection: Choose relevant few-shot examples based on the input query.
+- Fallback mechanism: Implement a fallback to standard querying if Self-Ask doesn't improve the result.
+- Result validation: Add a step to validate the final answer against the initial query.
+
+Here's an example of how you might implement dynamic example selection:
+
+```python
+from mirascope.core import openai, prompt_template
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics.pairwise import cosine_similarity
+import numpy as np
+
+
+def select_relevant_examples(
+    query: str, examples: list[FewShotExample], n: int = 3
+) -> list[FewShotExample]:
+    """Select the most relevant examples based on cosine similarity."""
+    vectorizer = TfidfVectorizer().fit([ex.question for ex in examples] + [query])
+    example_vectors = vectorizer.transform([ex.question for ex in examples])
+    query_vector = vectorizer.transform([query])
+
+    similarities = cosine_similarity(query_vector, example_vectors)[0]
+    most_similar_indices = np.argsort(similarities)[-n:][::-1]
+
+    return [examples[i] for i in most_similar_indices]
+
+
+@openai.call(model="gpt-4o-mini")
+@prompt_template(
+    """
+    MESSAGES: {example_prompts}
+    USER: {query}
+    """
+)
+def dynamic_self_ask_query(
+    query: str, examples: list[FewShotExample], n: int = 3
+) -> openai.OpenAIDynamicConfig:
+    relevant_examples = select_relevant_examples(query, examples, n)
+    return {
+        "computed_fields": {
+            "example_prompts": [
+                message
+                for example in relevant_examples
+                for message in example.messages()
+            ]
+        }
+    }
+
+
+query = "What was the primary language spoken by the inventor of the phonograph?"
```

Comment:
this answers my above question, but just confirming that the question asked to the model is, going forward, named query

Reply:
Yeah I think `query` is a good / common naming convention to use here going forward.

If you agree please also resolve the other comment you're mentioning.